pkg/controller/build/build_controller.go

```diff
@@ -235,67 +233,60 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 
 	// Populate the status struct with default values
 	b.Status.Registered = corev1.ConditionFalse
-	b.Status.Reason = succeedStatus
+	b.Status.Reason = build.SucceedStatus
 
 	// Validate if remote repository exists
 	if b.Spec.Source.SecretRef == nil {
 		if err := r.validateSourceURL(ctx, b, b.Spec.Source.URL); err != nil {
-			b.Status.Reason = err.Error()
-			if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
-				return reconcile.Result{}, updateErr
-			}
-			return reconcile.Result{}, nil
+			MarkBuildStatus(b, build.RemoteRepositoryUnreachable, err.Error())
+			return r.UpdateBuildStatusAndRetreat(ctx, b)
 		}
 	}
 
 	// Validate if the referenced secrets exist in the namespace
-	var secretNames []string
+	secretRefMap := map[string]build.BuildReason{}
```

Comment:
A *suggestion* (i.e. I do not feel adamant about it) ... let me work to it:
- so we already agreed to making the git URL validation optional via an annotation that could turn it off
- if over time, we have either this existing set of validations, or new ones that get created, we have some others that we may want to allow to be turned off ...
- we might want to have a more uniform means of building the list of validations, and iterating over them

To that end, perhaps:
- an array of validations, where maybe that array is built by a central utility method that looks at the annotation(s) we define for controlling validations
- where we define a golang interface for the defining the semantic of the validation
- and then this code becomes a simple for loop that iterates over each "validation" interface instance and calls `r.UpdateBuildStatusAndRetreat` as needed
- each validation's implementation of that interface is in a separate file from build_controller.go, so it is easy to gleam the set of validations via file names in a validation subfolder


Thoughts?

Reply:
hi @gabemontero , to summarize we have:
- 5 validations in total
- Only 1 validation is optional, the one of git URL. The rest are always triggered, there is no good reason not to do this afaik.

I do see the benefit on what you are saying and I would opt-in for such an enhancement, at least around:
- an external file with a well define interface for validations
- an approach to iterate on objects that implement this interface

However, for the goal we want to achieve, and based on what I mentioned above(only 1 validation is optional), I think this enhancement is out of the scope of the goal. I welcome you to provide this enhancement once this PR is merged.

Reply:
One could argue this is an implementation detail @qu1queee and not a new enhancement

WDYT @adambkaplan @sbose78 ?

Reply:
I agree with both of you @qu1queee and @gabemontero. Your proposal, @gabemontero, is a nice suggestion and will help us to make the controller file itself smaller by externalizing the validations. But, I also agree that this is not the intent of this PR. The build controller before this PR has methods like `validateRuntime`, `validationBuildStrategy` and `validateSecrets`. This design is maintained with the changes from @qu1queee although signatures had to be slightly changed to implement the requirement of representing the validation state of the build in a different way. The design change imo can be separated into another PR to keep also this PR and the one your propose smaller.

Reply:
Though one to be honest, I very much like the idea of reconciling all validations we have. Thing is that I do not know exactly how easy it would be to come up with an interface that suites all kind of validations we already have. Since this might a bit of time, I can understand to do that enhancement in another PR. However, I could also argue that now our understanding of this is fresh and we touch it anyway and things we want to do later can always delay and so on, that it would actually be an idea to do it now. Happy to help for what is worth with the implement if we decide to do it now.

Reply:
@gabemontero @HeavyWombat thanks, but I have to disagree on this one. This PR is intended to fix the following:

- #463 . This is the second PR around this, the first one was #483 which we closed. The goal here is to enhance the Build CRD status. Lets call this a "Build Status fields enhancement".
- #494 . Please read the issue, this is an enhancement around providing more clarity on existing validations failures on secrets. We are not introducing anything new, only making the failures more clear. Lets call this a "Existing Secrets validation failure enhancement". The code changes in the Reconcile logic you see, are because of this, which are improving/simplifying the code.

@gabemontero what you are proposing is:

- externalize the validations into a different package with a well defined interface that all validations implement. This should provide us an easier management of new validations in the "future".

As you could see, "Build Status fields enhancement" and "Existing Secrets validation failure enhancement" scope is not related to your proposal and there is no strong argument against doing this proposal as a follow up PR.

My recommendation is to move your proposal into a new issue with a proper description and acceptance criteria, so that is very clear on the intentions around what needs to be done. Getting this PR enhancements in place now,  will allow us to fix two issues(_these are coming users feedback_) and unblock us for shipping the v0.3.0 release.

Reply:
I may need to do a better job of explaining my suggestion @qu1queee ... but at the moment, I do not agree with your characterization of it.

I'll take another stab at illustrating it later today, as well as try to sanity check myself with @adambkaplan and/or @sbose78 to try and close on how to proceed.

But let's not merge this until that happens.

Reply:
@gabemontero of course, dont worry about this being merged, as long as you don´t approve it. Lets try to find consensus, see my proposal:

> My recommendation is to move your proposal into a new issue with a proper description and acceptance criteria, so that is very clear on the intentions around what needs to be done. Getting this PR enhancements in place now, will allow us to fix two issues(these are from users feedback) and unblock us for shipping the v0.3.0 release.

Reply:
I added one more commit with an implementation based on the discussions in here, im resolving this conversation and re-requesting reviews.